rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,453 @@
+ - tsh --proxy=proxy.example.com --user=dev headless approve <request_id>
+ - https://proxy.example.com/headless/<request_id>/approve
+// Wait for user to complete local authentication with MFA
+<user@node01> $
+```
+
+#### Environment variables
+
+In the `tsh --headless` flow, users never run `tsh login` on their remote machine. Instead, we expect the `--proxy`, `--user`, and `--headless` flags to be supplied to each command. To reduce UX friction, users can set the environment variables `TELEPORT_PROXY=<proxy_addr>`, `TELEPORT_USER=<user>`, and `TELEPORT_HEADLESS=true` instead.
+
+We prefer setting environment variables rather than saving config to disk (`~/.tsh/proxy.example.com.yaml` and `~/.tsh/current-profile`) so that the headless flow remains stateless, preventing conflicts on shared machines.
+
+#### `tsh headless`
+
+When the user enters the headless authentication command or URL, the user will be prompted to login with MFA, if they are not logged in already. The user will then be notified of additional request details and asked to acknowledge that they made this request (`y/N` prompt). Finally, the user is asked to verify with MFA to approve the request.
+
+```console
+$ tsh --proxy=proxy.example.com --user=dev headless approve <request_id>
+Enter password for Teleport user dev:
+Tap your YubiKey
+> Profile URL:        https://proxy.example.com
+  Logged in as:       dev
+  ...
+Headless request requires authentication. Contact your administrator if you didn't initiate this request.
+Additional details:
+  - command: "tsh ssh user@localhost"
+  - request id: <request_id>
+  - ip address: <ip_address>
+Approve? (y/N):
+Tap your YubiKey
+$ tsh headless accept <request_id> --proxy=proxy.example.com --user=dev
+Headless request requires authentication. Contact your administrator if you didn't initiate this request.
+Additional details:
+  - command: "tsh ssh user@localhost"
+  - request id: <request_id>
+  - ip address: <ip_address>
+Approve? (y/N):
+Tap your YubiKey
+```
+
+Note: When the user has to log in for the first time, we do not reuse their MFA verification to skip the second MFA check. Although this would be better UX, we cannot retrieve additional request details to share with the user until they log in. For security reasons, we should provide an MFA check after sharing the headless request details.
+
+#### Watch/View headless requests
+
+To improve UX, we will also offer an option for a user to watch for and accept headless requests. This will reduce the friction of copy-pasting the command or URL for every headless request.
+
+For `tsh`, we will add a new command, `tsh headless watch`, to watch for new requests. This will use the existing watcher API, filtering for the `HeadlessLoginRequest` resource for the user. When a request is found, its details will be shared in the user's local terminal. They will be prompted to acknowledge it (`y/N`) and then approve it with MFA.
```

Comment:
Is this really necessary? Feels like scope-creep a bit tbh. If we provide a way to approve/deny these requests in the web UI, that should cover most scenarios and provide better UX, no?

Reply:
Indeed, this will be removed.